File: src/SproAdapter/odbc.cs

using System;
using System.Collections.Generic;

namespace SocketProAdapter {
    namespace ClientSide {
        public class COdbc : CAsyncDBHandler {
            public const uint sidOdbc = SocketProAdapter.BaseServiceID.sidODBC; //asynchronous ODBC service id
            public COdbc()
                : base(sidOdbc) {
            }

            /// <summary>
            /// You may use the protected constructor when extending this class
            /// </summary>
            /// <param name="sid">A service id</param>
            protected COdbc(uint sid)
                : base(sid) {

            }

            //meta recordsets
            public const ushort idSQLColumnPrivileges = 0x7f00 + 100;
            public const ushort idSQLColumns = 0x7f00 + 101;
            public const ushort idSQLForeignKeys = 0x7f00 + 102;
            public const ushort idSQLPrimaryKeys = 0x7f00 + 103;
            public const ushort idSQLProcedureColumns = 0x7f00 + 104;
            public const ushort idSQLProcedures = 0x7f00 + 105;
            public const ushort idSQLSpecialColumns = 0x7f00 + 106;
            public const ushort idSQLStatistics = 0x7f00 + 107;
            public const ushort idSQLTablePrivileges = 0x7f00 + 108;
            public const ushort idSQLTables = 0x7f00 + 109;
            public const ushort idSQLGetInfo = 0x7f00 + 110;

            public const int ER_SUCCESS = 0;
            public const int ER_ERROR = -1;

            //error codes from async ODBC server library
            public const int ER_NO_DB_OPENED_YET = -1981;
            public const int ER_BAD_END_TRANSTACTION_PLAN = -1982;
            public const int ER_NO_PARAMETER_SPECIFIED = -1983;
            public const int ER_BAD_PARAMETER_COLUMN_SIZE = -1984;
            public const int ER_BAD_PARAMETER_DATA_ARRAY_SIZE = -1985;
            public const int ER_DATA_TYPE_NOT_SUPPORTED = -1986;
            public const int ER_NO_DB_NAME_SPECIFIED = -1987;
            public const int ER_ODBC_ENVIRONMENT_NOT_INITIALIZED = -1988;
            public const int ER_BAD_MANUAL_TRANSACTION_STATE = -1989;
            public const int ER_BAD_INPUT_PARAMETER_DATA_TYPE = -1990;
            public const int ER_BAD_PARAMETER_DIRECTION_TYPE = -1991;
            public const int ER_CORRECT_PARAMETER_INFO_NOT_PROVIDED_YET = -1992;

            private Dictionary<ushort, object> m_mapInfo = new Dictionary<ushort, object>();

            public object GetInfo(ushort infoType) {
                lock (m_csDB) {
                    if (m_mapInfo.ContainsKey(infoType))
                        return m_mapInfo[infoType];
                    return null;
                }
            }

            public bool ColumnPrivileges(string CatalogName, string SchemaName, string TableName, string ColumnName, DExecuteResult handler, DRows row, DRowsetHeader rh) {
                return DoMeta(idSQLColumnPrivileges, CatalogName, SchemaName, TableName, ColumnName, handler, row, rh, null);
            }

            public bool ColumnPrivileges(string CatalogName, string SchemaName, string TableName, string ColumnName, DExecuteResult handler, DRows row, DRowsetHeader rh, DDiscarded discarded) {
                return DoMeta(idSQLColumnPrivileges, CatalogName, SchemaName, TableName, ColumnName, handler, row, rh, discarded);
            }

            public bool Columns(string CatalogName, string SchemaName, string TableName, string ColumnName, DExecuteResult handler, DRows row, DRowsetHeader rh) {
                return DoMeta(idSQLColumns, CatalogName, SchemaName, TableName, ColumnName, handler, row, rh, null);
            }

            public bool Columns(string CatalogName, string SchemaName, string TableName, string ColumnName, DExecuteResult handler, DRows row, DRowsetHeader rh, DDiscarded discarded) {
                return DoMeta(idSQLColumns, CatalogName, SchemaName, TableName, ColumnName, handler, row, rh, discarded);
            }

            public bool ProcedureColumns(string CatalogName, string SchemaName, string ProcName, string ColumnName, DExecuteResult handler, DRows row, DRowsetHeader rh) {
                return DoMeta(idSQLProcedureColumns, CatalogName, SchemaName, ProcName, ColumnName, handler, row, rh, null);
            }

            public bool ProcedureColumns(string CatalogName, string SchemaName, string ProcName, string ColumnName, DExecuteResult handler, DRows row, DRowsetHeader rh, DDiscarded discarded) {
                return DoMeta(idSQLProcedureColumns, CatalogName, SchemaName, ProcName, ColumnName, handler, row, rh, discarded);
            }

            public bool PrimaryKeys(string CatalogName, string SchemaName, string TableName, DExecuteResult handler, DRows row, DRowsetHeader rh) {
                return DoMeta(idSQLPrimaryKeys, CatalogName, SchemaName, TableName, handler, row, rh, null);
            }

            public bool PrimaryKeys(string CatalogName, string SchemaName, string TableName, DExecuteResult handler, DRows row, DRowsetHeader rh, DDiscarded discarded) {
                return DoMeta(idSQLPrimaryKeys, CatalogName, SchemaName, TableName, handler, row, rh, discarded);
            }

            public bool TablePrivileges(string CatalogName, string SchemaName, string TableName, DExecuteResult handler, DRows row, DRowsetHeader rh) {
                return DoMeta(idSQLTablePrivileges, CatalogName, SchemaName, TableName, handler, row, rh, null);
            }

            public bool TablePrivileges(string CatalogName, string SchemaName, string TableName, DExecuteResult handler, DRows row, DRowsetHeader rh, DDiscarded discarded) {
                return DoMeta(idSQLTablePrivileges, CatalogName, SchemaName, TableName, handler, row, rh, discarded);
            }

            public bool Procedures(string CatalogName, string SchemaName, string ProcName, DExecuteResult handler, DRows row, DRowsetHeader rh) {
                return DoMeta(idSQLProcedures, CatalogName, SchemaName, ProcName, handler, row, rh, null);
            }

            public bool Procedures(string CatalogName, string SchemaName, string ProcName, DExecuteResult handler, DRows row, DRowsetHeader rh, DDiscarded discarded) {
                return DoMeta(idSQLProcedures, CatalogName, SchemaName, ProcName, handler, row, rh, discarded);
            }

            public bool SpecialColumns(short identifierType, string CatalogName, string SchemaName, string TableName, short scope, short nullable, DExecuteResult handler, DRows row, DRowsetHeader rh) {
                return DoMeta(idSQLSpecialColumns, identifierType, CatalogName, SchemaName, TableName, scope, nullable, handler, row, rh, null);
            }

            public bool SpecialColumns(short identifierType, string CatalogName, string SchemaName, string TableName, short scope, short nullable, DExecuteResult handler, DRows row, DRowsetHeader rh, DDiscarded discarded) {
                return DoMeta(idSQLSpecialColumns, identifierType, CatalogName, SchemaName, TableName, scope, nullable, handler, row, rh, discarded);
            }

            public bool Statistics(string CatalogName, string SchemaName, string TableName, ushort unique, ushort reserved, DExecuteResult handler, DRows row, DRowsetHeader rh) {
                return Statistics(CatalogName, SchemaName, TableName, unique, reserved, handler, row, rh, null);
            }

            public bool Statistics(string CatalogName, string SchemaName, string TableName, ushort unique, ushort reserved, DExecuteResult handler, DRows row, DRowsetHeader rh, DDiscarded discarded) {
                ulong index = GetCallIndex();
                //don't make m_csDB locked across calling SendRequest, which may lead to client dead-lock
                //in case a client asynchronously sends lots of requests without use of client side queue.
                lock (m_csDB) {
                    m_mapRowset[index] = new KeyValuePair<DRowsetHeader, DRows>(rh, row);
                }
                if (!SendRequest(idSQLStatistics, CatalogName, SchemaName, TableName, unique, reserved, index, (ar) => {
                    ProcessODBC(handler, ar, idSQLStatistics, index);
                }, discarded, null)) {
                    lock (m_csDB) {
                        m_mapRowset.Remove(index);
                    }
                    return false;
                }
                return true;
            }

            public bool Tables(string CatalogName, string SchemaName, string TableName, string TableType, DExecuteResult handler, DRows row, DRowsetHeader rh) {
                return DoMeta(idSQLTables, CatalogName, SchemaName, TableName, TableType, handler, row, rh, null);
            }

            public bool Tables(string CatalogName, string SchemaName, string TableName, string TableType, DExecuteResult handler, DRows row, DRowsetHeader rh, DDiscarded discarded) {
                return DoMeta(idSQLTables, CatalogName, SchemaName, TableName, TableType, handler, row, rh, discarded);
            }

            public bool ForeignKeys(string PKCatalogName, string PKSchemaName, string PKTableName, string FKCatalogName, string FKSchemaName, string FKTableName, DExecuteResult handler, DRows row, DRowsetHeader rh) {
                return DoMeta(idSQLForeignKeys, PKCatalogName, PKSchemaName, PKTableName, FKCatalogName, FKSchemaName, FKTableName, handler, row, rh, null);
            }

            public bool ForeignKeys(string PKCatalogName, string PKSchemaName, string PKTableName, string FKCatalogName, string FKSchemaName, string FKTableName, DExecuteResult handler, DRows row, DRowsetHeader rh, DDiscarded discarded) {
                return DoMeta(idSQLForeignKeys, PKCatalogName, PKSchemaName, PKTableName, FKCatalogName, FKSchemaName, FKTableName, handler, row, rh, discarded);
            }

            private void ProcessODBC(DExecuteResult handler, CAsyncResult ar, ushort reqId, ulong index) {
                ulong fail_ok;
                int res;
                string errMsg;
                ar.Load(out res).Load(out errMsg).Load(out fail_ok);
                COdbc odbc = (COdbc)ar.AsyncServiceHandler;
                lock (odbc.m_csDB) {
                    odbc.m_lastReqId = reqId;
                    odbc.m_affected = 0;
                    odbc.m_dbErrCode = res;
                    odbc.m_dbErrMsg = errMsg;
                    odbc.m_mapRowset.Remove(index);
                }
                if (handler != null)
                    handler(odbc, res, errMsg, 0, fail_ok, null);
            }

            private bool DoMeta(ushort id, string s0, string s1, string s2, string s3, DExecuteResult handler, DRows row, DRowsetHeader rh, DDiscarded discarded) {
                ulong index = GetCallIndex();
                //don't make m_csDB locked across calling SendRequest, which may lead to client dead-lock
                //in case a client asynchronously sends lots of requests without use of client side queue.
                lock (m_csDB) {
                    m_mapRowset[index] = new KeyValuePair<DRowsetHeader, DRows>(rh, row);
                }
                if (!SendRequest(id, s0, s1, s2, s3, index, (ar) => {
                    ProcessODBC(handler, ar, id, index);
                }, discarded, null)) {
                    lock (m_csDB) {
                        m_mapRowset.Remove(index);
                    }
                    return false;
                }
                return true;
            }

            private bool DoMeta(ushort id, string s0, string s1, string s2, DExecuteResult handler, DRows row, DRowsetHeader rh, DDiscarded discarded) {
                ulong index = GetCallIndex();
                //don't make m_csDB locked across calling SendRequest, which may lead to client dead-lock
                //in case a client asynchronously sends lots of requests without use of client side queue.
                lock (m_csDB) {
                    m_mapRowset[index] = new KeyValuePair<DRowsetHeader, DRows>(rh, row);
                }
                if (!SendRequest(id, s0, s1, s2, index, (ar) => {
                    ProcessODBC(handler, ar, id, index);
                }, discarded, null)) {
                    lock (m_csDB) {
                        m_mapRowset.Remove(index);
                    }
                    return false;
                }
                return true;
            }

            private bool DoMeta<T0, T1, T2>(ushort id, T0 t0, string s0, string s1, string s2, T1 t1, T2 t2, DExecuteResult handler, DRows row, DRowsetHeader rh, DDiscarded discarded) {
                ulong index = GetCallIndex();
                //don't make m_csDB locked across calling SendRequest, which may lead to client dead-lock
                //in case a client asynchronously sends lots of requests without use of client side queue.
                lock (m_csDB) {
                    m_mapRowset[index] = new KeyValuePair<DRowsetHeader, DRows>(rh, row);
                }
                if (!SendRequest(id, t0, s0, s1, s2, t1, t2, index, (ar) => {
                    ProcessODBC(handler, ar, id, index);
                }, discarded, null)) {
                    lock (m_csDB) {
                        m_mapRowset.Remove(index);
                    }
                    return false;
                }
                return true;
            }

            protected override void OnResultReturned(ushort reqId, CUQueue mc) {
                switch (reqId) {
                    case idSQLGetInfo:
                        lock (m_csDB) {
                            m_mapInfo.Clear();
                            while (mc.GetSize() > 0) {
                                ushort infoType;
                                object infoValue;
                                mc.Load(out infoType).Load(out infoValue);
                                m_mapInfo[infoType] = infoValue;
                            }
                        }
                        break;
                    default:
                        base.OnResultReturned(reqId, mc);
                        break;
                }
            }
        }
    }
}

File: bin/spa/clientside/odbc.py

from spa.memqueue import CUQueue, CScopeUQueue
from spa.clientside.asyncdbhandler import CAsyncDBHandler
from spa import BaseServiceID, tagBaseRequestID, Pair

class COdbc(CAsyncDBHandler):
    sidOdbc = BaseServiceID.sidODBC # asynchronous ODBC service id

    # meta recordsets
    idSQLColumnPrivileges = 0x7f00 + 100
    idSQLColumns = 0x7f00 + 101
    idSQLForeignKeys = 0x7f00 + 102
    idSQLPrimaryKeys = 0x7f00 + 103
    idSQLProcedureColumns = 0x7f00 + 104
    idSQLProcedures = 0x7f00 + 105
    idSQLSpecialColumns = 0x7f00 + 106
    idSQLStatistics = 0x7f00 + 107
    idSQLTablePrivileges = 0x7f00 + 108
    idSQLTables = 0x7f00 + 109
    idSQLGetInfo = 0x7f00 + 110;

    ER_SUCCESS = 0
    ER_ERROR = -1

    # error codes from async ODBC server library
    ER_NO_DB_OPENED_YET = -1981
    ER_BAD_END_TRANSTACTION_PLAN = -1982
    ER_NO_PARAMETER_SPECIFIED = -1983
    ER_BAD_PARAMETER_COLUMN_SIZE = -1984
    ER_BAD_PARAMETER_DATA_ARRAY_SIZE = -1985
    ER_DATA_TYPE_NOT_SUPPORTED = -1986
    ER_NO_DB_NAME_SPECIFIED = -1987
    ER_ODBC_ENVIRONMENT_NOT_INITIALIZED = -1988
    ER_BAD_MANUAL_TRANSACTION_STATE = -1989
    ER_BAD_INPUT_PARAMETER_DATA_TYPE = -1990
    ER_BAD_PARAMETER_DIRECTION_TYPE = -1991
    ER_CORRECT_PARAMETER_INFO_NOT_PROVIDED_YET = -1992

    def ColumnPrivileges(self, CatalogName, SchemaName, TableName, ColumnName, handler, row, rh, canceled = None):
        return self._DoMeta4(COdbc.idSQLColumnPrivileges, CatalogName, SchemaName, TableName, ColumnName, handler, row, rh, canceled)

    def Columns(self, CatalogName, SchemaName, TableName, ColumnName, handler, row, rh, canceled = None):
        return self._DoMeta4(COdbc.idSQLColumns, CatalogName, SchemaName, TableName, ColumnName, handler, row, rh, canceled)

    def ProcedureColumns(self, CatalogName, SchemaName, ProcName, ColumnName, handler, row, rh, canceled = None):
        return self._DoMeta4(COdbc.idSQLProcedureColumns, CatalogName, SchemaName, ProcName, ColumnName, handler, row, rh, canceled)

    def PrimaryKeys(self, CatalogName, SchemaName, TableName, handler, row, rh, canceled = None):
        return self._DoMeta3(COdbc.idSQLPrimaryKeys, CatalogName, SchemaName, TableName, handler, row, rh, canceled)

    def Procedures(self, CatalogName, SchemaName, ProcName, handler, row, rh, canceled = None):
        return self._DoMeta3(COdbc.idSQLProcedures, CatalogName, SchemaName, ProcName, handler, row, rh, canceled)

    def TablePrivileges(self, CatalogName, SchemaName, TableName, handler, row, rh, canceled = None):
        return self._DoMeta3(COdbc.idSQLTablePrivileges, CatalogName, SchemaName, TableName, handler, row, rh, canceled)

    def Tables(self, CatalogName, SchemaName, TableName, TableType, handler, row, rh, canceled = None):
        return self._DoMeta4(COdbc.idSQLTables, CatalogName, SchemaName, TableName, TableType, handler, row, rh, canceled)

    def _ProcessODBC_(self, handler, ar, reqId, index):
        mc = ar.UQueue;
        res = mc.LoadInt()
        errMsg = mc.LoadString()
        fail_ok = mc.LoadULong()
        with self._csDB:
            self._dbError = res
            self._dbErrorMsg = errMsg
            self._lastReqId = reqId
            self._affected = 0
            if index in self._mapRowset:
                self._mapRowset.pop(index)
        if handler:
            handler(self, res, errMsg, 0, fail_ok, None)

    def Statistics(self, CatalogName, SchemaName, TableName, unique, reserved, handler, row, rh, canceled = None):
        q = CScopeUQueue.Lock().SaveString(CatalogName).SaveString(SchemaName).SaveString(TableName).SaveUShort(unique).SaveUShort(reserved)
        ok = True
        with self._csOneSending:
            index = self.GetCallIndex()
            q.SaveULong(index)
            with self._csDB:
                self._mapRowset[index] = Pair(rh, row)
            ok = self.SendRequest(COdbc.idSQLStatistics, q, lambda ar: self._ProcessODBC_(handler, ar, COdbc.idSQLStatistics, index), canceled)
            if not ok:
                with self._csDB:
                    self._mapRowset.pop(index)
        CScopeUQueue.Unlock(q)
        return ok

    def SpecialColumns(self, identifierType, CatalogName, SchemaName, TableName, scope, nullable, handler, row, rh, canceled = None):
        q = CScopeUQueue.Lock().SaveShort(identifierType).SaveString(CatalogName).SaveString(SchemaName).SaveString(TableName).SaveShort(scope).SaveShort(nullable)
        ok = True
        with self._csOneSending:
            index = self.GetCallIndex()
            q.SaveULong(index)
            with self._csDB:
                self._mapRowset[index] = Pair(rh, row)
            ok = self.SendRequest(COdbc.idSQLSpecialColumns, q, lambda ar: self._ProcessODBC_(handler, ar, COdbc.idSQLSpecialColumns, index), canceled)
            if not ok:
                with self._csDB:
                    self._mapRowset.pop(index)
        CScopeUQueue.Unlock(q)
        return ok

    def ForeignKeys(self, PKCatalogName, PKSchemaName, PKTableName, FKCatalogName, FKSchemaName, FKTableName, handler, row, rh, canceled = None):
        q = CScopeUQueue.Lock().SaveString(PKCatalogName).SaveString(PKSchemaName).SaveString(PKTableName).SaveString(FKCatalogName).SaveString(FKSchemaName).SaveString(FKTableName)
        ok = True
        with self._csOneSending:
            index = self.GetCallIndex()
            q.SaveULong(index)
            with self._csDB:
                self._mapRowset[index] = Pair(rh, row)
            ok = self.SendRequest(COdbc.idSQLForeignKeys, q, lambda ar: self._ProcessODBC_(handler, ar, COdbc.idSQLForeignKeys, index), canceled)
            if not ok:
                with self._csDB:
                    self._mapRowset.pop(index)
        CScopeUQueue.Unlock(q)
        return ok

    def _DoMeta3(self, id, s0, s1, s2, handler, row, rh, canceled):
        q = CScopeUQueue.Lock().SaveString(s0).SaveString(s1).SaveString(s2)
        ok = True
        with self._csOneSending:
            index = self.GetCallIndex()
            q.SaveULong(index)
            with self._csDB:
                self._mapRowset[index] = Pair(rh, row)
            ok = self.SendRequest(id, q, lambda ar: self._ProcessODBC_(handler, ar, id, index), canceled)
            if not ok:
                with self._csDB:
                    self._mapRowset.pop(index)
        CScopeUQueue.Unlock(q)
        return ok

    def _DoMeta4(self, id, s0, s1, s2, s3, handler, row, rh, canceled):
        q = CScopeUQueue.Lock().SaveString(s0).SaveString(s1).SaveString(s2).SaveString(s3)
        ok = True
        with self._csOneSending:
            index = self.GetCallIndex()
            q.SaveULong(index)
            with self._csDB:
                self._mapRowset[index] = Pair(rh, row)
            ok = self.SendRequest(id, q, lambda ar: self._ProcessODBC_(handler, ar, id, index), canceled)
            if not ok:
                with self._csDB:
                    self._mapRowset.pop(index)
        CScopeUQueue.Unlock(q)
        return ok

    def __init__(self, sid=sidOdbc):
        super(COdbc, self).__init__(sid)
        self._mapInfo = {}

    def GetInfo(self, infoType):
        with self._csDB:
            if infoType in self._mapInfo:
                return self._mapInfo.get(infoType)

    def OnResultReturned(self, reqId, mc):
        if reqId == COdbc.idSQLGetInfo:
            with self._csDB:
                self._mapInfo = {}
                while mc.GetSize() > 0:
                    infoType = mc.LoadUShort()
                    vt = mc.LoadObject()
                    self._mapInfo[infoType] = vt
        else:
            super(COdbc, self).OnResultReturned(reqId, mc)


